File: python-algo/leetcode/problem_1110.py
"""1110. Delete Nodes And Return Forest
https://leetcode.com/problems/delete-nodes-and-return-forest/
"""
from typing import List, Optional

from common.tree_node import TreeNode


class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        def helper(node: Optional[TreeNode]) -> Optional[TreeNode]:
            if not node:
                return None
            node.left = helper(node.left)
            node.right = helper(node.right)
            if node.val not in del_nodes:
                return node
            if node.left:
                ans.append(node.left)
            if node.right:
                ans.append(node.right)
            return None

        ans = []
        del_nodes = set(to_delete)
        if helper(root):
            ans.append(root)
        return ans

File: java-algo/src/main/java/com/leetcode/Problem1110.java
package com.leetcode;

import com.common.TreeNode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * 1110. Delete Nodes And Return Forest
 * https://leetcode.com/problems/delete-nodes-and-return-forest/
 */
public class Problem1110 {
    Set<Integer> toDeleted = new HashSet<>();
    List<TreeNode> ans = new ArrayList<>();

    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        if (root == null) {
            return ans;
        }
        for (int d : to_delete) {
            toDeleted.add(d);
        }
        dfs(root, null, -1);
        return ans;
    }

    private void dfs(TreeNode cur, TreeNode par, int dir) {
        if (cur == null) {
            return;
        }
        if (toDeleted.isEmpty() && par != null) {
            return;
        }
        if (toDeleted.contains(cur.val)) {
            toDeleted.remove(cur.val);
            dfs(cur.left, null, -1);
            dfs(cur.right, null, -1);
            if (par != null) {
                if (dir == 0) {
                    par.left = null;
                } else {
                    par.right = null;
                }
            }
            return;
        }
        if (par == null) {
            ans.add(cur);
        }
        dfs(cur.left, cur, 0);
        dfs(cur.right, cur, 1);
    }
}
